session/pingpong/accountant_promise_handler.go

```diff
@@ -0,0 +1,320 @@
+}
+
+type encryption interface {
+	Decrypt(addr common.Address, encrypted []byte) ([]byte, error)
+	Encrypt(addr common.Address, plaintext []byte) ([]byte, error)
+}
+
+// AccountantPromiseHandlerDeps represents the AccountantPromiseHandler dependencies.
+type AccountantPromiseHandlerDeps struct {
+	AccountantPromiseStorage accountantPromiseStorage
+	AccountantCaller         accountantCaller
+	AccountantID             common.Address
+	FeeProvider              feeProvider
+	Encryption               encryption
+	EventBus                 eventbus.Publisher
+}
+
+// AccountantPromiseHandler handles the accountant promises for ongoing sessions.
+type AccountantPromiseHandler struct {
+	deps          AccountantPromiseHandlerDeps
+	queue         chan enqueuedRequest
+	stop          chan struct{}
+	stopOnce      sync.Once
+	startOnce     sync.Once
+	transactorFee registry.FeesResponse
+}
+
+// NewAccountantPromiseHandler returns a new instance of accountant promise handler.
+func NewAccountantPromiseHandler(deps AccountantPromiseHandlerDeps) *AccountantPromiseHandler {
+	return &AccountantPromiseHandler{
+		deps:  deps,
+		queue: make(chan enqueuedRequest, 100),
+		stop:  make(chan struct{}),
+	}
+}
+
+type enqueuedRequest struct {
+	errChan    chan error
+	r          []byte
+	em         crypto.ExchangeMessage
+	providerID identity.Identity
+	sessionID  string
+}
+
+// RequestPromise adds the request to the queue.
+func (aph *AccountantPromiseHandler) RequestPromise(r []byte, em crypto.ExchangeMessage, providerID identity.Identity, sessionID string) <-chan error {
+	er := enqueuedRequest{
+		r:          r,
+		em:         em,
+		providerID: providerID,
+		errChan:    make(chan error),
+		sessionID:  sessionID,
+	}
+	aph.queue <- er
+	return er.errChan
+}
+
+func (aph *AccountantPromiseHandler) updateFee() {
+	fees, err := aph.deps.FeeProvider.FetchSettleFees()
+	if err != nil {
+		log.Warn().Err(err).Msg("could not fetch fees, ignoring")
+		return
+	}
+
+	aph.transactorFee = fees
+}
+
+func (aph *AccountantPromiseHandler) handleRequests() {
+	log.Debug().Msgf("accountant promise handler started")
+	defer log.Debug().Msgf("accountant promise handler stopped")
+	for {
+		select {
+		case <-aph.stop:
+			return
+		case entry := <-aph.queue:
+			aph.requestPromise(entry)
+		}
+	}
+}
+
+// Subscribe subscribes AccountantPromiseHandler to relevant events.
+func (aph *AccountantPromiseHandler) Subscribe(bus eventbus.Subscriber) error {
+	err := bus.SubscribeAsync(event.AppTopicNode, aph.handleNodeStopEvents)
+	if err != nil {
+		return fmt.Errorf("could not subscribe to node events %w", err)
+	}
+
+	err = bus.SubscribeAsync(servicestate.AppTopicServiceStatus, aph.handleServiceEvent)
+	if err != nil {
+		return fmt.Errorf("could not subscribe to service events %w", err)
+	}
+	return nil
+}
+
+func (aph *AccountantPromiseHandler) handleServiceEvent(ev servicestate.AppEventServiceStatus) {
+	if ev.Status == string(servicestate.Running) {
+		aph.startOnce.Do(
```

Comment:
AccountantPromiseHandler is created as global instance once during node startup. If I'm correct:
start(ok)
stop(ok)
start(will not work as startOnce is used)
stop(stop will not work also)

Reply:
That's the intended behaviour. It is only stopped on node exit, and started once the first time we launch the service.